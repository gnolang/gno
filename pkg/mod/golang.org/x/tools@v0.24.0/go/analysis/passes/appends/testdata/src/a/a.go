// Copyright 2023 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// This file contains tests for the appends checker.

package a

func badAppendSlice1() {
	sli := []string{"a", "b", "c"}
	sli = append(sli) // want "append with no values"
}

func badAppendSlice2() {
	_ = append([]string{"a"}) // want "append with no values"
}

func goodAppendSlice1() {
	sli := []string{"a", "b", "c"}
	sli = append(sli, "d")
}

func goodAppendSlice2() {
	sli1 := []string{"a", "b", "c"}
	sli2 := []string{"d", "e", "f"}
	sli1 = append(sli1, sli2...)
}

func goodAppendSlice3() {
	sli := []string{"a", "b", "c"}
	sli = append(sli, "d", "e", "f")
}
