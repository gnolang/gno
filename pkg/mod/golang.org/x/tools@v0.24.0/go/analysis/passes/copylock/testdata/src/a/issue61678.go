package a

import "sync"

// These examples are taken from golang/go#61678, modified so that A and B
// contain a mutex.

type A struct {
	a  A
	mu sync.Mutex
}

type B struct {
	a  A
	b  B
	mu sync.Mutex
}

func okay(x A) {}
func sure()    { var x A; nop(x) }

var fine B

func what(x B)   {}                  // want `passes lock by value`
func bad()       { var x B; nop(x) } // want `copies lock value`
func good()      { nop(B{}) }
func stillgood() { nop(B{b: B{b: B{b: B{}}}}) }
func nope()      { nop(B{}.b) } // want `copies lock value`

func nop(any) {} // only used to get around unused variable errors
