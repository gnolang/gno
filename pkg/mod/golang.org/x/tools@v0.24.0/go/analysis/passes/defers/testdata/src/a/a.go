// Copyright 2023 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package a

import (
	"fmt"
	"time"
)

func Since() (t time.Duration) {
	return
}

func x(time.Duration) {}
func x2(float64)      {}

func good() {
	// The following are OK because func is not evaluated in defer invocation.
	now := time.Now()
	defer func() {
		fmt.Println(time.Since(now)) // OK because time.Since is not evaluated in defer
	}()
	evalBefore := time.Since(now)
	defer fmt.Println(evalBefore)
	do := func(f func()) {}
	defer do(func() { time.Since(now) })
	defer fmt.Println(Since())       // OK because Since function is not in module time
	defer copy([]int(nil), []int{1}) // check that a builtin doesn't cause a panic
}

type y struct{}

func (y) A(float64)        {}
func (*y) B(float64)       {}
func (y) C(time.Duration)  {}
func (*y) D(time.Duration) {}

func bad() {
	var zero time.Time
	now := time.Now()
	defer time.Since(zero)                    // want "call to time.Since is not deferred"
	defer time.Since(now)                     // want "call to time.Since is not deferred"
	defer fmt.Println(time.Since(now))        // want "call to time.Since is not deferred"
	defer fmt.Println(time.Since(time.Now())) // want "call to time.Since is not deferred"
	defer x(time.Since(now))                  // want "call to time.Since is not deferred"
	defer x2(time.Since(now).Seconds())       // want "call to time.Since is not deferred"
	defer y{}.A(time.Since(now).Seconds())    // want "call to time.Since is not deferred"
	defer (&y{}).B(time.Since(now).Seconds()) // want "call to time.Since is not deferred"
	defer y{}.C(time.Since(now))              // want "call to time.Since is not deferred"
	defer (&y{}).D(time.Since(now))           // want "call to time.Since is not deferred"
}

func ugly() {
	// The following is ok even though time.Since is evaluated. We don't
	// walk into function literals or check what function definitions are doing.
	defer x((func() time.Duration { return time.Since(time.Now()) })())
}
