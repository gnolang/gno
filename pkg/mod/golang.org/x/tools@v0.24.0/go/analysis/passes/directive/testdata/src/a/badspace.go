// Copyright 2023 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

//go:build ignore

// want +1 `invalid space '\\u00a0' in //go:debug directive`
//go:debugÂ 00a0

package main

