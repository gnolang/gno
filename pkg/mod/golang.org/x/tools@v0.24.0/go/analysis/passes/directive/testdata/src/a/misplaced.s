// Copyright 2023 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// want +1 `//go:debug directive only valid in Go source files`
//go:debug panicnil=1

/*
can skip over comments
//go:debug doesn't matter here
*/

// want +1 `//go:debug directive only valid in Go source files`
//go:debug panicnil=1

package a

// no error here because we can't parse this far
//go:debug panicnil=1
