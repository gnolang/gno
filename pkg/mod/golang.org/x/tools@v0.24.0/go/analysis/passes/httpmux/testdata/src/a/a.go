// Copyright 2023 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// This file contains tests for the httpmux checker.

package a

import "net/http"

func _() {
	http.HandleFunc("GET /x", nil)  // want "enhanced ServeMux pattern"
	http.HandleFunc("/{a}/b/", nil) // want "enhanced ServeMux pattern"
	mux := http.NewServeMux()
	mux.Handle("example.com/c/{d}", nil) // want "enhanced ServeMux pattern"
	mux.HandleFunc("/{x...}", nil)       // want "enhanced ServeMux pattern"

	// Should not match.

	// not an enhanced pattern
	http.Handle("/", nil)

	// invalid wildcard; will panic in 1.22
	http.HandleFunc("/{/a/}", nil)
	mux.Handle("/{1}", nil)
	mux.Handle("/x{a}", nil)

	// right package, wrong method
	http.ParseTime("GET /")

	// right function name, wrong package
	Handle("GET /", nil)
	HandleFunc("GET /", nil)

	// right method name, wrong type
	var sm ServeMux
	sm.Handle("example.com/c/{d}", nil)
	sm.HandleFunc("method /{x...}", nil)
}

func Handle(pat string, x any)     {}
func HandleFunc(pat string, x any) {}

type ServeMux struct{}

func (*ServeMux) Handle(pat string, x any)     {}
func (*ServeMux) HandleFunc(pat string, x any) {}
