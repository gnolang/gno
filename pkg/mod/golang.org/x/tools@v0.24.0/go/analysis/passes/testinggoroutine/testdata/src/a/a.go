// Copyright 2020 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package a

import (
	"log"
	"sync"
	"testing"
)

func TestBadFatalf(t *testing.T) {
	var wg sync.WaitGroup
	defer wg.Wait()

	for i := 0; i < 2; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			t.Fatalf("TestFailed: id = %v\n", id) // want "call to .+T.+Fatalf from a non-test goroutine"
		}(i)
	}
}

func TestOKErrorf(t *testing.T) {
	var wg sync.WaitGroup
	defer wg.Wait()

	for i := 0; i < 2; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			t.Errorf("TestFailed: id = %v\n", id)
		}(i)
	}
}

func TestBadFatal(t *testing.T) {
	var wg sync.WaitGroup
	defer wg.Wait()

	for i := 0; i < 2; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			t.Fatal("TestFailed") // want "call to .+T.+Fatal from a non-test goroutine"
		}(i)
	}
}

func f(t *testing.T, _ string) {
	t.Fatal("TestFailed")
}

func g() {}

func TestBadFatalIssue47470(t *testing.T) {
	go f(t, "failed test 1") // want "call to .+T.+Fatal from a non-test goroutine"

	g := func(t *testing.T, _ string) {
		t.Fatal("TestFailed")
	}
	go g(t, "failed test 2") // want "call to .+T.+Fatal from a non-test goroutine"
}

func BenchmarkBadFatalf(b *testing.B) {
	var wg sync.WaitGroup
	defer wg.Wait()

	for i := 0; i < b.N; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			b.Fatalf("TestFailed: id = %v\n", id) // want "call to .+B.+Fatalf from a non-test goroutine"
		}(i)
	}
}

func BenchmarkBadFatal(b *testing.B) {
	var wg sync.WaitGroup
	defer wg.Wait()

	for i := 0; i < b.N; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			b.Fatal("TestFailed") // want "call to .+B.+Fatal from a non-test goroutine"
		}(i)
	}
}

func BenchmarkOKErrorf(b *testing.B) {
	var wg sync.WaitGroup
	defer wg.Wait()

	for i := 0; i < b.N; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			b.Errorf("TestFailed: %d", i)
		}(i)
	}
}

func BenchmarkBadFatalGoGo(b *testing.B) {
	var wg sync.WaitGroup
	defer wg.Wait()

	for i := 0; i < b.N; i++ {
		wg.Add(1)
		go func(id int) {
			go func() {
				defer wg.Done()
				b.Fatal("TestFailed") // want "call to .+B.+Fatal from a non-test goroutine"
			}()
		}(i)
	}

	if false {
		defer b.Fatal("here")
	}

	if true {
		go func() {
			b.Fatal("in here") // want "call to .+B.+Fatal from a non-test goroutine"
		}()
	}

	func() {
		func() {
			func() {
				func() {
					go func() {
						b.Fatal("Here") // want "call to .+B.+Fatal from a non-test goroutine"
					}()
				}()
			}()
		}()
	}()

	_ = 10 * 10
	_ = func() bool {
		go b.Fatal("Failed") // want "call to .+B.+Fatal from a non-test goroutine"
		return true
	}

	defer func() {
		go b.Fatal("Here") // want "call to .+B.+Fatal from a non-test goroutine"
	}()
}

func BenchmarkBadSkip(b *testing.B) {
	for i := 0; i < b.N; i++ {
		if i == 100 {
			go b.Skip("Skipping") // want "call to .+B.+Skip from a non-test goroutine"
		}
		if i == 22 {
			go func() {
				go func() {
					b.Skip("Skipping now") // want "call to .+B.+Skip from a non-test goroutine"
				}()
			}()
		}
	}
}

func TestBadSkip(t *testing.T) {
	for i := 0; i < 1000; i++ {
		if i == 100 {
			go t.Skip("Skipping") // want "call to .+T.+Skip from a non-test goroutine"
		}
		if i == 22 {
			go func() {
				go func() {
					t.Skip("Skipping now") // want "call to .+T.+Skip from a non-test goroutine"
				}()
			}()
		}
	}
}

func BenchmarkBadFailNow(b *testing.B) {
	for i := 0; i < b.N; i++ {
		if i == 100 {
			go b.FailNow() // want "call to .+B.+FailNow from a non-test goroutine"
		}
		if i == 22 {
			go func() {
				go func() {
					b.FailNow() // want "call to .+B.+FailNow from a non-test goroutine"
				}()
			}()
		}
	}
}

func TestBadFailNow(t *testing.T) {
	for i := 0; i < 1000; i++ {
		if i == 100 {
			go t.FailNow() // want "call to .+T.+FailNow from a non-test goroutine"
		}
		if i == 22 {
			go func() {
				go func() {
					t.FailNow() // want "call to .+T.+FailNow from a non-test goroutine"
				}()
			}()
		}
	}
}

func TestBadWithLoopCond(ty *testing.T) {
	var wg sync.WaitGroup
	defer wg.Wait()

	for i := 0; i < 10; i++ {
		wg.Add(1)
		go func(id int) {
			defer ty.Fatalf("Why") // want "call to .+T.+Fatalf from a non-test goroutine"
			go func() {
				for j := 0; j < 2; ty.FailNow() { // want "call to .+T.+FailNow from"
					j++
					ty.Errorf("Done here")
				}
			}()
		}(i)
	}
}

type customType int

func (ct *customType) Fatalf(fmtSpec string, args ...interface{}) {
	if fmtSpec == "" {
		panic("empty format specifier")
	}
}

func (ct *customType) FailNow() {}
func (ct *customType) Skip()    {}

func TestWithLogFatalf(t *testing.T) {
	var wg sync.WaitGroup
	defer wg.Wait()

	for i := 0; i < 10; i++ {
		wg.Add(1)
		go func(id int) {
			go func() {
				for j := 0; j < 2; j++ {
					log.Fatal("Done here")
				}
			}()
		}(i)
	}
}

func TestWithCustomType(t *testing.T) {
	var wg sync.WaitGroup
	defer wg.Wait()

	ct := new(customType)
	defer ct.FailNow()
	defer ct.Skip()

	for i := 0; i < 10; i++ {
		wg.Add(1)
		go func(id int) {
			go func() {
				for j := 0; j < 2; j++ {
					ct.Fatalf("Done here: %d", i)
				}
			}()
		}(i)
	}
}

func helpTB(tb testing.TB) {
	tb.FailNow()
}

func TestTB(t *testing.T) {
	go helpTB(t) // want "call to .+TB.+FailNow from a non-test goroutine"
}

func TestIssue48124(t *testing.T) {
	go helper(t) // want "call to .+T.+Skip from a non-test goroutine"
}

func TestEachCall(t *testing.T) {
	go helper(t) // want "call to .+T.+Skip from a non-test goroutine"
	go helper(t) // want "call to .+T.+Skip from a non-test goroutine"
}

func TestWithSubtest(t *testing.T) {
	t.Run("name", func(t2 *testing.T) {
		t.FailNow() // want "call to .+T.+FailNow on t defined outside of the subtest"
		t2.Fatal()
	})

	f := func(t3 *testing.T) {
		t.FailNow()
		t3.Fatal()
	}
	t.Run("name", f) // want "call to .+T.+FailNow on t defined outside of the subtest"

	g := func(t4 *testing.T) {
		t.FailNow()
		t4.Fatal()
	}
	g(t)

	t.Run("name", helper)

	go t.Run("name", func(t2 *testing.T) {
		t.FailNow() // want "call to .+T.+FailNow on t defined outside of the subtest"
		t2.Fatal()
	})
}

func TestMultipleVariables(t *testing.T) {
	{ // short decl
		f, g := func(t1 *testing.T) {
			t1.Fatal()
		}, func(t2 *testing.T) {
			t2.Error()
		}

		go f(t) // want "call to .+T.+Fatal from a non-test goroutine"
		go g(t)

		t.Run("name", f)
		t.Run("name", g)
	}

	{ // var decl
		var f, g = func(t1 *testing.T) {
			t1.Fatal()
		}, func(t2 *testing.T) {
			t2.Error()
		}

		go f(t) // want "call to .+T.+Fatal from a non-test goroutine"
		go g(t)

		t.Run("name", f)
		t.Run("name", g)
	}
}

func BadIgnoresMultipleAssignments(t *testing.T) {
	{
		f := func(t1 *testing.T) {
			t1.Fatal()
		}
		go f(t) // want "call to .+T.+Fatal from a non-test goroutine"

		f = func(t2 *testing.T) {
			t2.Error()
		}
		go f(t) // want "call to .+T.+Fatal from a non-test goroutine"
	}
	{
		f := func(t1 *testing.T) {
			t1.Error()
		}
		go f(t)

		f = func(t2 *testing.T) {
			t2.FailNow()
		}
		go f(t) // false negative
	}
}

func TestGoDoesNotDescendIntoSubtest(t *testing.T) {
	f := func(t2 *testing.T) {
		g := func(t3 *testing.T) {
			t3.Fatal() // fine
		}
		t2.Run("name", g)
		t2.FailNow() // bad
	}
	go f(t) // want "call to .+T.+FailNow from a non-test goroutine"
}

func TestFreeVariableAssignedWithinEnclosing(t *testing.T) {
	f := func(t2 *testing.T) {
		inner := t
		inner.FailNow()
	}

	go f(nil) // want "call to .+T.+FailNow from a non-test goroutine"

	t.Run("name", func(t3 *testing.T) {
		go f(nil) // want "call to .+T.+FailNow from a non-test goroutine"
	})

	// Without pointer analysis we cannot tell if inner is t or t2.
	// So we accept a false negatives on the following examples.
	t.Run("name", f)

	go func(_ *testing.T) {
		t.Run("name", f)
	}(nil)

	go t.Run("name", f)
}

func TestWithUnusedSelection(t *testing.T) {
	go func() {
		_ = t.FailNow
	}()
	t.Run("name", func(t2 *testing.T) {
		_ = t.FailNow
	})
}

func TestMethodExprsAreIgnored(t *testing.T) {
	go func() {
		(*testing.T).FailNow(t)
	}()
}

func TestRecursive(t *testing.T) {
	t.SkipNow()

	go TestRecursive(t) // want "call to .+T.+SkipNow from a non-test goroutine"

	t.Run("name", TestRecursive)
}

func TestMethodSelection(t *testing.T) {
	var h helperType

	go h.help(t) // want "call to .+T.+SkipNow from a non-test goroutine"
	t.Run("name", h.help)
}

type helperType struct{}

func (h *helperType) help(t *testing.T) { t.SkipNow() }

func TestIssue63799a(t *testing.T) {
	done := make(chan struct{})
	go func() {
		defer close(done)
		t.Run("", func(t *testing.T) {
			t.Fatal() // No warning. This is in a subtest.
		})
	}()
	<-done
}

func TestIssue63799b(t *testing.T) {
	// Simplified from go.dev/cl/538698

	// nondet is some unspecified boolean placeholder.
	var nondet func() bool

	t.Run("nohup", func(t *testing.T) {
		if nondet() {
			t.Skip("ignored")
		}

		go t.Run("nohup-i", func(t *testing.T) {
			t.Parallel()
			if nondet() {
				if nondet() {
					t.Skip("go.dev/cl/538698 wanted to have skip here")
				}

				t.Error("ignored")
			} else {
				t.Log("ignored")
			}
		})
	})
}

func TestIssue63849(t *testing.T) {
	go func() {
		helper(t) // False negative. We do not do an actual interprodecural reachability analysis.
	}()
	go helper(t) // want "call to .+T.+Skip from a non-test goroutine"
}
