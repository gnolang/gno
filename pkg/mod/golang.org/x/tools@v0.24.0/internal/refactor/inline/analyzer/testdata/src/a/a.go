package a

func f() {
	One() // want `inline call of a.One`

	new(T).Two() // want `inline call of \(a.T\).Two`
}

type T struct{}

// inlineme
func One() int { return one } // want One:`inlineme a.One`

const one = 1

// inlineme
func (T) Two() int { return 2 } // want Two:`inlineme \(a.T\).Two`
