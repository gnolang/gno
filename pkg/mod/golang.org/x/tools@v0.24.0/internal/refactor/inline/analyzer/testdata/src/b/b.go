package b

import "a"

func f() {
	a.One() // want `cannot inline call to a.One because body refers to non-exported one`

	new(a.T).Two() // want `inline call of \(a.T\).Two`
}
