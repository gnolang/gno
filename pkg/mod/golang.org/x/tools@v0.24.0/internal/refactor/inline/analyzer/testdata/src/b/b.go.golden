package b

import "a"

func f() {
	a.One() // want `cannot inline call to a.One because body refers to non-exported one`

	_ = 2 // want `inline call of \(a.T\).Two`
}
